pkg/steps/source.go

```diff
@@ -539,20 +542,51 @@ func hintsAtInfraReason(logSnippet string) bool {
 		strings.Contains(logSnippet, "connection reset by peer")
 }
 
-func waitForBuildOrTimeout(ctx context.Context, buildClient BuildClient, namespace, name string) error {
-	return waitForBuild(ctx, buildClient, namespace, name)
+func waitForBuildOrTimeout(
+	ctx context.Context,
+	buildClient BuildClient,
+	podClient kubernetes.PodClient,
+	namespace, name string,
+) error {
+	return waitForBuild(ctx, buildClient, podClient, namespace, name)
 }
 
-func waitForBuild(ctx context.Context, buildClient BuildClient, namespace, name string) error {
+func waitForBuild(
+	ctx context.Context,
+	buildClient BuildClient,
+	podClient kubernetes.PodClient,
+	namespace, name string,
+) error {
 	logrus.WithFields(logrus.Fields{
 		"namespace": namespace,
 		"name":      name,
 	}).Trace("Waiting for build to be complete.")
-
-	evaluatorFunc := func(obj runtime.Object) (bool, error) {
-		switch build := obj.(type) {
-		case *buildapi.Build:
+	var ret atomic.Pointer[buildapi.Build]
```

Comment:
> The use we make of this pointer is tricky (nasty?):

The (atomic) pointer provides the latest version of the object as seen by the watch.

> The whole logic is probably correct but is non-trivial. Wouldn't a channel be more appropriate in this case where you are trying to make two goroutines communicate?

We could use a channel, but I don't see how that makes things clearer.
- it would just be a send/receive instead of a store/load
- it would require an extra boolean variable to know when to start the pending check
- there is only ever one load from the pointer, so the channel would be single-use

Other than those reasons, this is analogous to the same pattern in the function linked above which watches pods, where the pointer makes things even simpler.  I did consider using a channel there, feeding both the watch and the pending check events through it, and processing both streams uniformly.  However, those two events are unrelated: we know exactly when and what needs to be performed for the pending check (in both functions), and mixing the treatment in this way makes no sense.

> the *buildapi.Build object here could be different from the following line if in the meantime a swap happened due to a change of status.

As desired.  The first load is used to determine the point at which the verification needs to be performed, based on the creation timestamp.  The second (which happens at a much later point in time) has to perform another load since it must consider the latest version of the object.  As mentioned in the description, not doing this was one of the problems in the original implementation.

Reply:
I have tried to quickly modify the code using a channel and it turned out to be even more convoluted so, again (see the comment below), nether I am in favor to go down this path anymore.
The main problem was that I had to rebuild your mental model of the function (see [here](https://raw.githubusercontent.com/zakirullin/cognitive-load/main/img/cognitiveloadv2.png) from [Cognitive Load Developer's Handbook](https://github.com/zakirullin/cognitive-load?utm_source=programmingdigest&utm_medium&utm_campaign=1648)), which carry on a lot of hidden details and assumptions, so I was trying to reduce a little bit the complexity starting by eliminating the communication through the atomic pointer.
Comments were probably the real missing piece, I can now "see" the what you intended and feel more relaxed.